### ISSUE TITLE
Performance drop on small models trained on CPU between 0.3.1 and 0.4

### ISSUE THREAD

**Issue by jhmenke** [pytorch/pytorch#8154]
## Issue description

Updating from Pytorch 0.3.1 to 0.4 yields a significant performance drop on small feedforward networks, trained **on CPU**. This effect happens on Linux (in a virtual machine) and on Windows.

The attached code was executed on a Windows 10 machine with an Intel i7-7700HQ.
The Linux test was done in a virtual machine. The computer is equipped with an Intel i7-4702MQ.

The average runtime over 10 repetitions on Linux is 26.141655 s for Pytorch 0.3.1 and 32.092158 s on 0.4  (~23 % increase)
The average runtime over 10 repetitions on Windows is 12.577887 s for Pytorch 0.3.1 and 17.759720 s on 0.4  (~41 % increase)

Observations: 
- The time difference between 0.3.1 and 0.4 seems to be more or less constant on both Windows and Linux. 
- Also profiling suggests the DataLoader __next__ takes 8 s instead of 4 s previously, a major part of the increased time
- Weirdly, in different calls of the script, the early stopping if-clause activates sometimes despire setting the pytorch seed as to avoid effects due to randomness

## Code example

See my results and the data/script used here:
[PytorchSpeedTest.zip](https://github.com/pytorch/pytorch/files/2071136/PytorchSpeedTest.zip)

## System Info

(the collect_env is not yet adapted to Windows, therefore information is missing. I am using pip 10.x and the latest conda)

PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.1

OS: Microsoft Windows 10 Home
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): all conda (Windows 0.3.1. via conda install -c peterjc123 pytorch-cpu)
- OS: Windows 10 / Ubuntu 16.04 
- PyTorch version: 0.3.1 to 0.4
- Python version: 3.6

## Bottleneck Profiling
0.4
```
--------------------------------------------------------------------------------
  cProfile output
--------------------------------------------------------------------------------
         12550526 function calls (12252338 primitive calls) in 21.593 seconds

   Ordered by: internal time
   List reduced from 334 to 15 due to restriction <15>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    31780    4.498    0.000    4.498    0.000 {method 'run_backward' of 'torch._C._EngineBase' objects}
    84840    2.774    0.000    2.774    0.000 {built-in method stack}
  3049200    2.219    0.000    2.219    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\utils\data\dataset.py:40(<genexpr>)
    31780    1.347    0.000    3.899    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\optim\adam.py:48(step)
   127260    1.217    0.000    1.217    0.000 {built-in method addmm}
  1016400    0.866    0.000    3.085    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\utils\data\dataset.py:39(__getitem__)
   190680    0.593    0.000    0.593    0.000 {method 'addcdiv_' of 'torch._C._TensorBase' objects}
296940/84840    0.566    0.000    3.832    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\nn\modules\module.py:485(__call__)
   381360    0.557    0.000    0.557    0.000 {method 'mul_' of 'torch._C._TensorBase' objects}
    84840    0.553    0.000    0.553    0.000 {built-in method torch._C._nn.threshold}
    42560    0.532    0.000    8.012    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\utils\data\dataloader.py:261(__next__)
        1    0.511    0.511   21.593   21.593 .\pytorch.0.4.py:1(<module>)
   381360    0.470    0.000    0.470    0.000 {method 'add_' of 'torch._C._TensorBase' objects}
    42560    0.438    0.000    0.603    0.000 C:\Users\Jan\Miniconda3\lib\site-packages\torch\utils\data\sampler.py:136(__iter__)
   190680    0.421    0.000    0.421    0.000 {method 'sqrt' of 'torch._C._TensorBase' objects}


--------------------------------------------------------------------------------
  autograd profiler output (CPU mode)
--------------------------------------------------------------------------------
        top 15 events sorted by cpu_time_total

----------------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                           CPU time        CUDA time            Calls        CPU total       CUDA total
----------------------  ---------------  ---------------  ---------------  ---------------  ---------------
stack                         475.169us          0.000us                1        475.169us          0.000us
stack                         415.362us          0.000us                1        415.362us          0.000us
stack                         388.741us          0.000us                1        388.741us          0.000us
stack                         381.812us          0.000us                1        381.812us          0.000us
stack                         378.531us          0.000us                1        378.531us          0.000us
stack                         371.966us          0.000us                1        371.966us          0.000us
ThresholdBackward0            370.507us          0.000us                1        370.507us          0.000us
threshold_backward            369.049us          0.000us                1        369.049us          0.000us
AddmmBackward                 358.473us          0.000us                1        358.473us          0.000us
stack                         350.815us          0.000us                1        350.815us          0.000us
mm                            347.169us          0.000us                1        347.169us          0.000us
select                        343.887us          0.000us                1        343.887us          0.000us
add_                          340.239us          0.000us                1        340.239us          0.000us
unsqueeze                     339.511us          0.000us                1        339.511us          0.000us
addmm                         335.864us          0.000us                1        335.864us          0.000us
```

### cProfile
0.3.1

 ``` 
 12819090 function calls (12546756 primitive calls) in 14.661 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    28602    3.038    0.000    3.038    0.000 {method 'run_backward' of 'torch._C._EngineBase' objects}
  1829520    1.041    0.000    1.041    0.000 {method 'unsqueeze' of 'torch._C.FloatTensorBase' objects}
    28602    0.855    0.000    2.765    0.000 adam.py:31(step)
   114534    0.854    0.000    0.854    0.000 {built-in method torch._C.addmm}
   914760    0.616    0.000    0.616    0.000 dataset.py:42(__getitem__)
   171612    0.504    0.000    0.504    0.000 {method 'sqrt' of 'torch._C.FloatTensorBase' objects}
   171612    0.468    0.000    0.468    0.000 {method 'addcdiv_' of 'torch._C.FloatTensorBase' objects}
    38304    0.466    0.000    0.917    0.000 sampler.py:117(__iter__)
        1    0.454    0.454   14.661   14.661 pytorch.0.3.1.py:1(<module>)
267246/76356    0.438    0.000    3.045    0.000 module.py:351(__call__)
   343224    0.437    0.000    0.437    0.000 {method 'mul_' of 'torch._C.FloatTensorBase' objects}
    76356    0.431    0.000    0.431    0.000 {built-in method torch._C._nn.threshold}
114534/38178    0.421    0.000    2.340    0.000 dataloader.py:99(default_collate)
    38304    0.272    0.000    4.413    0.000 dataloader.py:256(__next__)
    76356    0.270    0.000    0.270    0.000 {built-in method torch._C.cat}
```

**Comment by cpuhrsch** [comment]
Thank you for reporting this separately @jhmenke.

**Comment by cpuhrsch** [comment]
cc @peterjc123

**Comment by peterjc123** [comment]
@jhmenke @cpuhrsch  Well, I don't think it is Windows specific. Although I have changed `THAllocator` that may be related to this, it should not be the case according to your test results. The difference in time seems constant, so it must be somewhere else.

**Comment by jhmenke** [comment]
Since the Linux version also shows decreased performance i also don't think there is any connection to the build process of peterjc123.
I personally am unclear how i would debug this further..

**Comment by jhmenke** [comment]
This issue could be related to #8255 
I will test the performance if i find a quiet hour this week.

**Comment by jhmenke** [comment]
It is now way worse:

I just cloned and compiled PyTorch (torch-0.5.0a0+e31ab99, e31ab99) on Linux. v0.31 average was ~26 s, v0.4 average ~ 32 s. 
Now the average time is: 45.49 s

torch.bottleneck runs and after a few minutes exits with the message "killed" without giving any output. here is the manual cProfile log:

`12393230 function calls (12099688 primitive calls) in 42.208 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    30872    7.610    0.000    7.610    0.000 {method 'run_backward' of 'torch._C._EngineBase' objects}
    82416    4.467    0.000    4.467    0.000 {built-in method stack}
  2962080    4.061    0.000    4.061    0.000 dataset.py:40(<genexpr>)
   123624    3.209    0.000    3.209    0.000 {built-in method addmm}
    30872    2.495    0.000    7.598    0.000 adam.py:48(step)
   987360    1.904    0.000    5.964    0.000 dataset.py:39(__getitem__)
    82416    1.306    0.000    1.306    0.000 {built-in method torch._C._nn.threshold}
288456/82416    1.256    0.000    8.898    0.000 module.py:462(__call__)
   185232    1.136    0.000    1.136    0.000 {method 'sqrt' of 'torch._C._TensorBase' objects}
   370464    1.106    0.000    1.106    0.000 {method 'add_' of 'torch._C._TensorBase' objects}
        1    1.047    1.047   42.208   42.208 pytorch_debug.0.4.py:1(<module>)
    41208    0.989    0.000    6.953    0.000 dataloader.py:314(<listcomp>)
   370464    0.965    0.000    0.965    0.000 {method 'mul_' of 'torch._C._TensorBase' objects}
    41344    0.958    0.000   14.878    0.000 dataloader.py:311(__next__)
   185232    0.937    0.000    0.937    0.000 {method 'addcdiv_' of 'torch._C._TensorBase' objects}
    41344    0.927    0.000    1.252    0.000 sampler.py:136(__iter__)
   185232    0.667    0.000    0.667    0.000 {method 'addcmul_' of 'torch._C._TensorBase' objects}
   123624    0.620    0.000    4.994    0.000 linear.py:54(forward)
   123624    0.604    0.000    0.604    0.000 {method 't' of 'torch._C._TensorBase' objects}
123624/41208    0.557    0.000    5.679    0.000 dataloader.py:151(default_collate)
    41208    0.441    0.000    8.015    0.000 container.py:89(forward)
    41208    0.403    0.000    0.403    0.000 {built-in method torch._C._nn.smooth_l1_loss}
   185226    0.369    0.000    0.369    0.000 {method 'zero_' of 'torch._C._TensorBase' objects}
    30872    0.324    0.000    0.734    0.000 optimizer.py:150(zero_grad)
   123624    0.308    0.000    4.171    0.000 functional.py:995(linear)
    30872    0.256    0.000    0.256    0.000 {built-in method ones_like}
1497619/1497435    0.248    0.000    0.249    0.000 {built-in method builtins.len}`

**Comment by cpuhrsch** [comment]
We've recently added a few performance improvements. Could you try again @jhmenke?

**Comment by jhmenke** [comment]
Yes, we are back at the 0.4 average:
>Average time: 31.492700 s

**Comment by cpuhrsch** [comment]
@jhmenke - Can we consider this issue resolved?

**Comment by jhmenke** [comment]
i guess so, thanks for your time!
